### PR TITLE
Kernel/FileSystem: Make Inode::{write,read}_bytes methods non-virtual

### PR DESCRIPTION
**Pull request by supercomputer7**
We make these methods non-virtual because we want to ensure we properly
enforce locking of the m_inode_lock mutex. Also, for write operations,
we want to call prepare_to_write_data before the actual write. The
previous design required us to ensure the callers do that at various
places which lead to hard-to-find bugs. By moving everything to a place
where we call prepare_to_write_data only once, we eliminate a possibilty
of forgeting to call it on some code path in the kernel.

cc @IdanHo

### CODE REVIEW

File: Kernel/FileSystem/Inode.cpp [hunk 1]
```diff
@@ -106,6 +106,19 @@ void Inode::will_be_destroyed()
...
+ErrorOr<size_t> Inode::write_bytes(off_t offset, size_t length, UserOrKernelBuffer const& target_buffer, OpenFileDescription* open_description)
+{
+    MutexLocker locker(m_inode_lock);
+    TRY(prepare_to_write_data());
+    return write_bytes_locked(offset, length, target_buffer, open_description);
+}
+
+ErrorOr<size_t> Inode::read_bytes(off_t offset, size_t length, UserOrKernelBuffer& buffer, OpenFileDescription* open_description) const
+{
+    MutexLocker locker(m_inode_lock, Mutex::Mode::Shared);
```
Comment (IdanHo) [inline Kernel/FileSystem/Inode.cpp:118]: This doesn't seem safe. Some of our read_bytes implementations assume exclusive locking (e.g. Ext2FS)

**Review by IdanHo — COMMENTED** [review]
One comment, but looks good.
Also, can we add another commit that adds a Kernel test for shared mmaping a file, writing to it, adn then doing an msync on it, to ensure we don't accidentally break SharedInodeVMObjects in the future?

File: Kernel/FileSystem/Ext2FileSystem.cpp [hunk 2]
```diff
@@ -1499,9 +1499,9 @@ ErrorOr<NonnullRefPtr<Inode>> Ext2FS::create_inode(Ext2FSInode& parent_inode, St
     return new_inode;
 }
 
-ErrorOr<void> Ext2FSInode::populate_lookup_cache() const
+ErrorOr<void> Ext2FSInode::populate_lookup_cache()
 {
-    MutexLocker locker(m_inode_lock);
+    MutexLocker locker(m_lookup_cache_lock);
```
Comment (IdanHo) [inline Kernel/FileSystem/Ext2FileSystem.cpp:1516]: There's no need for this lock, all callers already hold the `m_inode_lock` exclusively, instead assert that `m_inode_lock` is held.

File: Kernel/FileSystem/Ext2FileSystem.cpp [hunk 3]
```diff
@@ -798,6 +798,19 @@ ErrorOr<NonnullRefPtr<Inode>> Ext2FS::get_inode(InodeIdentifier inode) const
...
+ErrorOr<void> Ext2FSInode::compute_block_list_with_exclusive_locking()
+{
+    MutexLocker block_list_locker(m_block_list_lock);
+    if (m_block_list.is_empty())
+        m_block_list = TRY(compute_block_list());
+
+    if (m_block_list.is_empty()) {
+        dmesgln("Ext2FSInode[{}]::read_bytes(): Empty block list", identifier());
+        return EIO;
+    }
```
Comment (IdanHo) [inline Kernel/FileSystem/Ext2FileSystem.cpp:810]: Why is this here? it doesn't really have anything to do with computing the block list, so it should be part of read_bytes.

File: Kernel/FileSystem/Ext2FileSystem.cpp (hunk 2 above)
Comment (supercomputer7) [reply Kernel/FileSystem/Ext2FileSystem.cpp:1516]: This is not true, because `Ext2FSInode::lookup` will try to use this method and as far as I can tell doesn't hold that lock when calling this method. We can change that but really I don't see why we should change the order of taking locks in the lookup method.

File: Kernel/FileSystem/Ext2FileSystem.cpp [hunk 4]
```diff
@@ -798,6 +798,14 @@ ErrorOr<NonnullLockRefPtr<Inode>> Ext2FS::get_inode(InodeIdentifier inode) const
     return new_inode;
 }
 
+ErrorOr<void> Ext2FSInode::compute_block_list_with_exclusive_locking()
+{
+    MutexLocker block_list_locker(m_block_list_lock);
```
Comment (IdanHo) [inline Kernel/FileSystem/Ext2FileSystem.cpp:803]: Can we add a comment explaining this messy lock situation? i.e., why is it legal that some holders only lock `m_inode_lock` before accessing `m_block_list`, while others only hold `m_inode_lock`? And while we're doing that, maybe also add `VERIFY` that `m_inode_lock` is being held in a shared state, not in an exclusive state, to confirm said new comment remains true?

File: Kernel/Memory/PrivateInodeVMObject.cpp [hunk 5]
```diff
@@ -11,7 +11,19 @@ namespace Kernel::Memory {
...
 {
-    auto new_physical_pages = TRY(VMObject::try_create_physical_pages(inode.size()));
+    if (inode.size() == 0)
+        return EINVAL;
+    return try_create_with_inode_and_range(inode, 0, inode.size());
+}
+
+ErrorOr<NonnullLockRefPtr<PrivateInodeVMObject>> PrivateInodeVMObject::try_create_with_inode_and_range(Inode& inode, u64 offset, size_t range_size)
+{
+    // FIXME: To ensure further allocation of a Region with this VMObject will not complain
```
Comment (IdanHo) [inline Kernel/Memory/PrivateInodeVMObject.cpp:21]: why is this a FIXME?

File: Kernel/Memory/PrivateInodeVMObject.cpp [hunk 6]
```diff
@@ -11,7 +11,19 @@ namespace Kernel::Memory {
...
+    // on "smaller" VMObject than the requested Region, we simply take the max size between both values.
+    auto size = max(inode.size(), (offset + range_size));
+    VERIFY(size > 0);
+    VERIFY(size > 0);
```
Comment (IdanHo) [inline Kernel/Memory/PrivateInodeVMObject.cpp:25]: double VERIFY

File: Kernel/Arch/x86/common/Interrupts.cpp [hunk 7]
```diff
@@ -317,14 +317,19 @@ void page_fault_handler(TrapFrame* trap)
...
-    if (response == PageFaultResponse::ShouldCrash || response == PageFaultResponse::OutOfMemory) {
+    if (response == PageFaultResponse::ShouldCrash || response == PageFaultResponse::OutOfMemory || response == PageFaultResponse::BusError) {
         if (faulted_in_kernel && handle_safe_access_fault(regs, fault_address)) {
             // If this would be a ring0 (kernel) fault and the fault was triggered by
             // safe_memcpy, safe_strnlen, or safe_memset then we resume execution at
             // the appropriate _fault label rather than crashing
             return;
         }
 
+        if (response == PageFaultResponse::BusError && current_thread->has_signal_handler(SIGBUS)) {
```
Comment (IdanHo) [inline Kernel/Arch/x86/common/Interrupts.cpp:328]: what if the current thread does not have a signal handler `SIGBUS`? the code below will crash it a SIGSEGV instead of a SIGBUS as the cause, that needs to be updated as well.

File: Kernel/Memory/SharedInodeVMObject.cpp [hunk 8]
```diff
@@ -12,7 +12,17 @@ namespace Kernel::Memory {
...
 {
-    size_t size = inode.size();
+    if (inode.size() == 0)
+        return EINVAL;
+    return try_create_with_inode_and_range(inode, 0, inode.size());
+}
+
+ErrorOr<NonnullLockRefPtr<SharedInodeVMObject>> SharedInodeVMObject::try_create_with_inode_and_range(Inode& inode, u64 offset, size_t range_size)
+{
+    // FIXME: To ensure further allocation of a Region with this VMObject will not complain
```
Comment (IdanHo) [inline Kernel/Memory/SharedInodeVMObject.cpp:22]: ditto

**Review by IdanHo — COMMENTED** [review]
Sorry, had a chance to re-review it properly again and found a couple extra small issues

**Review by IdanHo — APPROVED** [review]
Very nice work, will merge once CI passes
